src/dataflow/src/source/postgres.rs

```diff
@@ -0,0 +1,319 @@
+        let mut stream = client
+            .start_logical_replication(&slot_name, cur_lsn, &self.plugin)
+            .await
+            .unwrap();
+
+        let mut time_lease = None;
+        while let Some(item) = stream.next().await {
+            match item {
+                Ok(ReplicationMessage::XLogData(xlog_data)) => {
+                    use LogicalReplicationMessage::*;
+
+                    match xlog_data.data() {
+                        Begin(_) => {
+                            // Lease a tx time
+                            time_lease = Some(timestamper.lease().await);
+                        }
+                        Insert(insert) if insert.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let row = self.row_from_tuple(insert.tuple(), &col_types);
+                            co.yield_(Ok((row, time, 1))).await;
+                        }
+                        Update(update) if update.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let old_row =
+                                self.row_from_tuple(update.old_tuple().unwrap(), &col_types);
+                            co.yield_(Ok((old_row, time, -1))).await;
+
+                            let new_row = self.row_from_tuple(update.new_tuple(), &col_types);
+                            co.yield_(Ok((new_row, time, 1))).await;
+                        }
+                        Delete(delete) if delete.rel_id() == rel_id => {
+                            let time = time_lease
+                                .as_deref()
+                                .cloned()
+                                .expect("got row event without a prior BEGIN event");
+                            let row = self.row_from_tuple(delete.old_tuple().unwrap(), &col_types);
+                            co.yield_(Ok((row, time, -1))).await;
+                        }
+                        Commit(commit) => {
+                            cur_lsn = commit.commit_lsn().into();
+                            // Release the lease
+                            time_lease = None;
+                        }
+                        Origin(_) | Relation(_) | Type(_) => (),
+                        Truncate(_) => panic!("source table got truncated"), // TODO turn this into a stream error
+                        _ => panic!("unexpected logical replication message"),
+                    }
+                }
+                Ok(ReplicationMessage::PrimaryKeepAlive(keepalive)) => {
+                    if keepalive.reply() == 1 {
```

Comment:
I think I get what the rest of this function is doing but I have no idea what this block is doing so I think it could use a comment or two (the rest could use more comments too but this jumps out)

Reply:
I'll make a pass through this file to comment it but what this block does is sends feedback back to the postgres instance letting it know how far along we are in consuming the WAL.

This basically allows upstream to know when to cleanup WAL entries on disk. This has the nice sideeffect that you can monitor materialize replication lag with native pg replication monitoring tools

Reply:
Added a whole bunch of comments in that file